src/server/game/Entities/Player/Player.cpp

```diff
@@ -14074,8 +14074,15 @@ void Player::PrepareGossipMenu(WorldObject* source, uint32 menuId /*= 0*/, bool
                 case GOSSIP_OPTION_QUESTGIVER:
                     canTalk = false;
                     break;
-                case GOSSIP_OPTION_GOSSIP:
                 case GOSSIP_OPTION_TRAINER:
+                    if (getClass() != sObjectMgr->GetTrainer(creature->GetEntry())->GetTrainerRequirement() && sObjectMgr->GetTrainer(creature->GetEntry())->GetTrainerType() == Trainer::Type::Class)
```

Comment:
Are we only denying the wrong class here? Or should wrong race/missing spell also be checked here?

Reply:
no idea. armor/weaponsmiting was removed on retail, only profession trainers are left

Reply:
Well, race needs checking for riding trainers, so you have got a point there.